simulation_test.go

```diff
@@ -0,0 +1,169 @@
+
+import (
+	"crypto/sha256"
+	"io/ioutil"
+	"testing"
+
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+	"github.com/tidwall/gjson"
+)
+
+const nidSize = 2
+
+func getLeavesFromState(state gjson.Result, nidSize int) [][]byte {
+	itf_leaves := state.Get("leaves_v").Array()
+	var leaves [][]byte
+
+	for _, itf_leaf := range itf_leaves {
+		stateNamespace := int(itf_leaf.Get("namespaceId").Int())
+		dataString := itf_leaf.Get("value.#tup.1").String()
+		pushData := namespace.PrefixedData(append(
+			intToBytes(stateNamespace, nidSize),
+			[]byte(dataString)...))
+		leaves = append(leaves, pushData)
+
+	}
+	return leaves
+}
+
+// checking if the proof generated by the execution matches the proof in the model
+// hashes are not checked because the model is using a simplified hash function
+func checkProof(modelProof gjson.Result, execProof nmt.Proof, nidSize int) bool {
+	modelProofStart := int(modelProof.Get("start").Int())
+	execProofStart := execProof.Start()
+
+	modelProofEnd := int(modelProof.Get("end").Int())
+	execProofEnd := execProof.End()
+	if modelProofStart != execProofStart || modelProofEnd != execProofEnd {
+		return false
+	}
+	supporting_hashes := modelProof.Get("supporting_hashes").Array()
+	if len(supporting_hashes) != len(execProof.Nodes()) {
+		return false
+	}
+	for idx, supporting_hash := range supporting_hashes {
+		modelMinNS := int(supporting_hash.Get("minNS").Int())
+		execMinNs := bytesToInt(execProof.Nodes()[idx][:nidSize])
+		if modelMinNS != execMinNs {
+			return false
+		}
+
+		modelMaxNS := int(supporting_hash.Get("maxNS").Int())
+		execMaxNs := bytesToInt(execProof.Nodes()[idx][nidSize:(2 * nidSize)])
+		if modelMaxNS != execMaxNs {
+			return false
+		}
+	}
+
+	return true
+
+}
+
+func intToBytes(n int, bytesSize int) []byte {
+	b := make([]byte, bytesSize)
+	for i := 0; i < bytesSize; i++ {
+		b[i] = byte(n >> (8 * (bytesSize - i - 1)))
+	}
+	return b
+}
+
+func bytesToInt(bytes []byte) int {
+	var result int
+	for i, b := range bytes {
+		shift := uint((len(bytes) - 1 - i) * 8)
+		result |= int(b) << shift
+	}
+	return result
+}
+
+func TestFromITF(t *testing.T) {
+
+	var tree *nmt.NamespacedMerkleTree
+
+	itfFileName := "formal_spec/ITF_files/runTest.itf.json"
+	// itfFileName := "formal_spec/ITF_files/panicIssue.itf.json"
+	data, err := ioutil.ReadFile(itfFileName)
+	if err != nil {
+		t.Errorf("Error opening file: %v", err)
+	}
+	states := gjson.GetBytes(data, "states").Array()
+
+	// iterate over all states of the test run
+	for stateIdx, state := range states {
+		t.Logf("State %v\n", stateIdx)
+		// because the test run consists of multiple steps to generate the tree and the proof, test
+		// only the states marked "final"
+		if state.Get("state_v").String() == "final" {
+			modelNamespace := int(state.Get("namespace_v").Int())
+			modelLeaves := getLeavesFromState(state, nidSize)
+			modelProof := state.Get("proof_v")
+			corrupted := state.Get("corrupted").Bool()
```

Comment:
After reviewing the provided descriptions, it appears that the nmtTest module generates only corrupted proofs. If this is the case, then why is the corrupted boolean being checked in this line? Shouldn't it always be true?

Reply:
It is actually not generating only corrupted proofs.

If you look at the definition of the `step` action, it says:
```bluespec
    action step = {
        any{

            // 1): requirements: defining the tree and the namespace to generate a proof for
            requirements,

            // 2): generation of the proof 
            generation,
            
            // 2b): corrupting the proof
            corruptProof,            

            // 3): collecting all results
            final
            
        }

```

The `any` keyword means that the action to run is chosen non-deterministically among the four option. We do restrict this non-determinism by conditions, e.g., after `requirements`, only `generation` can happen. However, after `generation`, either `corruptProof` or `final` may happen. In the latter case, we'll have a non-corrupted proof.

This thing (that both `corruptProof` and `final` may happen) is defined by their preconditions: both of them require `state_v == "final"`: action `corruptProof` at [line 718](https://github.com/ivan-gavran/nmt/blob/c3cc6b7acba34c97a1a4d5e8fa4be1d355535c1e/formal_spec/nmt.qnt#LL718C14-L718C14) and action `final` at [line 695](https://github.com/ivan-gavran/nmt/blob/c3cc6b7acba34c97a1a4d5e8fa4be1d355535c1e/formal_spec/nmt.qnt#L695).